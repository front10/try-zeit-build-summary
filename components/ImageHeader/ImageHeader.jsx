/* eslint-disable prettier/prettier */
import React from "react";
import PropTypes from "prop-types";
import useStyles from "@front10/helpers/dist/UseStyles";
import Box from "../Box";
import Text from "../Text";
import {
  isElementInViewport,
  loadImage,
  getScrollParent
} from "../../helpers/imageLazyLoading";

class ImageHeader extends React.Component {
  constructor(props) {
    super(props);
    this.loaded = false;
    this.image = null;
    this.scrollParents = [];
    this.id = `fr-image-2`;
  }

  componentDidUpdate = prevProps => {
    this.onUpdate(prevProps);
  };

  handleLoadImage = () => {
    const { image, imageHeight, imageWidth, pathNoImageFound, imageBgPos } = this.props;
    this.image = loadImage(
      this.loaded,
      this.image,
      image,
      imageHeight,
      imageWidth,
      pathNoImageFound,
      imageBgPos
    );
  };

  onUpdate = prevProps => {
    const { image, imageHeight, imageWidth } = this.props;
    if (
      image !== prevProps.image ||
      (image === prevProps.image && this.loaded === false)
    ) {
      this.loaded = false;
      if (this.image) {
        this.image.setAttribute(
          "style",
          `background-image:none;height:${imageHeight};width:${imageWidth};backgroundColor:lightgray;animation:pulse 1s infinite ease-in-out`
        );
        if (isElementInViewport(this.id)) this.handleLoadImage();
      }
    }
  };

  componentDidMount = () => {
    if (this.image) {
      setTimeout(() => {
        this.scrollParents = getScrollParent(this.image, this.scrollParents);
        if (isElementInViewport(this.id)) {
          this.handleLoadImage();
        } else this.addEventListener();
      });
    }
  };

  componentWillUnmount = () => {
    this.removeEventListener();
  };

  listener = () => {
    clearTimeout(this.scrollTimmer);
    this.scrollTimmer = setTimeout(() => {
      if (isElementInViewport(this.id) && !this.loaded) this.handleLoadImage();
    }, 250);
  };

  addEventListener = () => {
    this.scrollParents.map(node => {
      if (node && typeof node.addEventListener === "function") {
        node.addEventListener("scroll", this.listener, true);
      }
      return node;
    });
  };

  removeEventListener = () => {
    this.scrollParents.map(node => {
      if (node && typeof node.addEventListener === "function")
        node.removeEventListener("scroll", this.listener, true);
      return node;
    });
  };

  noLabelImage = () => {
    const { noLabelImageSettings } = this.props;
    return (
      <Box className="cmp-image-label" display="none">
        <Text {...noLabelImageSettings}>Image not found</Text>
      </Box>
    );
  };

  render = () => {
    const {
      imageHeight,
      imageWidth,
      hoverEffect,
      alt,
      focusable,
      children
    } = this.props;
    const [className, style] = useStyles(this.props);
    const loadingImageStyle = {
      ...style,
      height: imageHeight,
      width: imageWidth,
      backgroundColor: "lightgray",
      animation: "pulse 1s infinite ease-in-out"
    };
    return (
      <Box
        id={this.id}
        tabIndex={focusable ? 0 : null}
        focusable={focusable}
        height="100"
        className={`fr-${ImageHeader.componentName}-container`}
        role="img"
        dataTestId="imageHeader"
        ariaLabel={alt}
      >
        <Box
          className={`fr-image-header ${hoverEffect &&
            `fr-effects fr-effect-${hoverEffect}`} ${className}`}
          style={loadingImageStyle}
          focusReferences={imageLoadedElem => {
            this.image = imageLoadedElem;
          }}
        >
          {children}
          {this.noLabelImage()}
        </Box>
      </Box>
    );
  };
}

ImageHeader.propTypes = {
  /**
   * URI of the image.
   */
  image: PropTypes.string,
  /**
   * alt description
   */
  alt: PropTypes.string,
  /**
   * Height of Destination Image
   */
  imageHeight: PropTypes.string,
  /**
   * Width of Destination Image
   */
  imageWidth: PropTypes.string,
  /**
   * Effect apply of the image
   */
  hoverEffect: PropTypes.oneOf([
    "grayscale",
    "fadeIn",
    "scale",
    "blur",
    "opacity",
    "sepia",
    "lightly",
    "dark"
  ]),
  /**
   * Enable focusable in this components.
   */
  focusable: PropTypes.bool,
  /**
   *  Children of the component
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  pathNoImageFound: PropTypes.string,
  noLabelImageSettings: PropTypes.objectOf(PropTypes.any)
};
ImageHeader.defaultProps = {
  image: "",
  alt: null,
  imageHeight: "250px",
  imageWidth: "100%",
  hoverEffect: null,
  focusable: true,
  children: null,
  pathNoImageFound:
    "https://res.cloudinary.com/dakp804eh/image/upload/v1561045947/airmodules/Utils/NoImageFound.svg",
  noLabelImageSettings: {
    position: "absolute",
    padding: "3",
    bottom: "0",
    right: "0"
  },
  imageBgPos: "inherit"
};

ImageHeader.componentName = "ImageHeader";

export default ImageHeader;
