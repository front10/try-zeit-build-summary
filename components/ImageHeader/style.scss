.fr-ImageHeader-container {
  position: relative;
  .fr-image-header {
    background-repeat: no-repeat;
    background-size: cover;
    min-width: 100%;
    background-position: center;
  }

  .fr-effects {
    filter: none;
    transition: 400ms ease;
  }
  &:hover,
  &:focus {
    .fr-ImageHeader-overlay {
      opacity: 1;
    }
    .fr-effects {
      transition: 400ms ease;
    }
    // Filters applied on the images
    .fr-effect-grayscale {
      filter: grayscale(100%);
      -webkit-filter: grayscale(100%);
    }
    .fr-effect-fadeIn {
      opacity: 0.7;
    }
    .fr-effect-scale {
      transform: scale(1.03);
      box-shadow: 0px 0px 9px 3px rgba(0, 0, 0, 0.2);
    }
    .fr-effect-blur {
      filter: blur(2px);
    }
    .fr-effect-opacity {
      filter: brightness(50%);
      -webkit-filter: brightness(50%);
    }
    .fr-effect-sepia {
      filter: sepia(100%);
      -webkit-filter: sepia(100%);
    }
    .fr-effect-lightly {
      filter: brightness(130%);
      -webkit-filter: brightness(130%);
    }
    .fr-effect-dark {
      filter: brightness(85%);
      -webkit-filter: brightness(85%);
    }
  }
  .fr-ImageHeader-overlay {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    height: 100%;
    width: 100%;
    opacity: 1;
    transition: 0.5s ease;
    background-color: rgba(0, 0, 0, 0.15);
    .fr-ImageHeader-text {
      color: white;
      font-size: 20px;
      position: absolute;
      top: 50%;
      left: 50%;
      text-shadow: 2px 2px 2px black;
      width: 100%;
      -webkit-transform: translate(-50%, -50%);
      -ms-transform: translate(-50%, -50%);
      transform: translate(-50%, -50%);
      text-align: center;
    }
  }
}
