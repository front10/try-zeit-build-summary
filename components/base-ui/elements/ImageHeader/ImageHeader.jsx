/* eslint-disable prettier/prettier */
import React from 'react';
import PropTypes from 'prop-types';
import { getUuid } from '@front10/helpers/dist/uuid';
import useStyles from '@front10/helpers/dist/UseStyles';
import Box from '../Box';
import Text from '../Text';
import {
  isElementInViewport,
  loadImage,
  getScrollParent
} from '../../helpers/imageLazyLoading';

class ImageHeader extends React.Component {
  constructor(props) {
    super(props);
    this.loaded = false;
    this.image = null;
    this.scrollParents = [];
    this.id = `fr-image-${getUuid()}`;
  }

  componentDidUpdate = prevProps => {
    this.onUpdate(prevProps);
  };

  handleLoadImage = () => {
    const { image, imageHeight, imageWidth, pathNoImageFound } = this.props;
    this.image = loadImage(
      this.loaded,
      this.image,
      image,
      imageHeight,
      imageWidth,
      pathNoImageFound
    );
  };

  onUpdate = prevProps => {
    const { image, imageHeight, imageWidth } = this.props;
    if (image !== prevProps.image || (image === prevProps.image && this.loaded === false)) {
      this.loaded = false;
      if (this.image) {
        this.image.setAttribute(
          'style',
          `background-image:none;height:${imageHeight};width:${imageWidth};backgroundColor:lightgray;animation:pulse 1s infinite ease-in-out`
        );
        if (isElementInViewport(this.id)) this.handleLoadImage();
      }
    }
  };

  componentDidMount = () => {
    if (this.image) {
      setTimeout(() => {
        this.scrollParents = getScrollParent(this.image, this.scrollParents);
        if (isElementInViewport(this.id)) {
          this.handleLoadImage();
        } else this.addEventListener();
      });
    }
  };

  componentWillUnmount = () => {
    this.removeEventListener();
  };

  listener = () => {
    clearTimeout(this.scrollTimmer);
    this.scrollTimmer = setTimeout(() => {
      if (isElementInViewport(this.id) && !this.loaded) this.handleLoadImage();
    }, 250);
  };

  addEventListener = () => {
    this.scrollParents.map(node => {
      if (node && typeof node.addEventListener === 'function') {
        node.addEventListener('scroll', this.listener, true);
      }
      return node;
    });
  };

  removeEventListener = () => {
    this.scrollParents.map(node => {
      if (node && typeof node.addEventListener === 'function')
        node.removeEventListener('scroll', this.listener, true);
      return node;
    });
  };

  noLabelImage = () => {
    const { translate, noLabelImageSettings } = this.props;
    return (
      <Box className="cmp-image-label" display="none">
        <Text {...noLabelImageSettings}>{translate('cmp-image-header-label-not-image-found')}</Text>
      </Box>
    );
  };

  render = () => {
    const { imageHeight, imageWidth, hoverEffect, alt, focusable, children } = this.props;
    const [className, style] = useStyles(this.props);
    const loadingImageStyle = {
      ...style,
      height: imageHeight,
      width: imageWidth,
      backgroundColor: 'lightgray',
      animation: 'pulse 1s infinite ease-in-out'
    };
    return (
      <Box
        id={this.id}
        tabIndex={focusable ? 0 : null}
        focusable={focusable}
        height="100"
        className={`fr-${ImageHeader.componentName}-container`}
        role="img"
        dataTestId="imageHeader"
        ariaLabel={alt}
      >
        <Box
          className={`fr-image-header ${hoverEffect &&
            `fr-effects fr-effect-${hoverEffect}`} ${className}`}
          style={loadingImageStyle}
          focusReferences={imageLoadedElem => {
            this.image = imageLoadedElem;
          }}
        >
          {children}
          {this.noLabelImage()}
        </Box>
      </Box>
    );
  };
}

ImageHeader.propTypes = {
  /**
   * URI of the image.
   */
  image: PropTypes.string,
  /**
   * alt description
   */
  alt: PropTypes.string,
  /**
   * Height of Destination Image
   */
  imageHeight: PropTypes.string,
  /**
   * Width of Destination Image
   */
  imageWidth: PropTypes.string,
  /**
   * Effect apply of the image
   */
  hoverEffect: PropTypes.oneOf([
    'grayscale',
    'fadeIn',
    'scale',
    'blur',
    'opacity',
    'sepia',
    'lightly',
    'dark'
  ]),
  /**
   * Enable focusable in this components.
   */
  focusable: PropTypes.bool,
  /**
   *  Children of the component
   */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  pathNoImageFound: PropTypes.string,
  translate: PropTypes.func,
  noLabelImageSettings: PropTypes.objectOf(PropTypes.any)
};
ImageHeader.defaultProps = {
  image: '',
  alt: null,
  imageHeight: '250px',
  imageWidth: '100%',
  hoverEffect: null,
  focusable: true,
  children: null,
  pathNoImageFound:
    'https://res.cloudinary.com/dakp804eh/image/upload/v1561045947/airmodules/Utils/NoImageFound.svg',
  translate: () => {},
  noLabelImageSettings: {
    position: 'absolute',
    padding: '3',
    bottom: '0',
    right: '0'
  }
};

ImageHeader.componentName = 'ImageHeader';

export default ImageHeader;
