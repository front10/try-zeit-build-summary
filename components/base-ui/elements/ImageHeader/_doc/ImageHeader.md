# ImageHeader
  [![standard-readme compliant](https://img.shields.io/badge/standard--readme-OK-green.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)
  

  ## Properties
  | </br>Name | </br>Type | </br>Summary | </br>Default | 
| ---- | ---- | ---- | ---- |
| alt | `string` | alt description | `null` |
| children | `union` | Children of the component | `null` |
| focusable | `bool` | Enable focusable in this components. | `true` |
| hoverEffect | `enum` | Effect apply of the image | `null` |
| image | `string` | URI of the image. | `''` |
| imageHeight | `string` | Height of Destination Image | `'250px'` |
| imageWidth | `string` | Width of Destination Image | `'100%'` |
| noLabelImageSettings | `objectOf` |  | `{  position: 'absolute',  padding: '3',  bottom: '0',  right: '0'}` |
| pathNoImageFound | `string` |  | `'https://res.cloudinary.com/dakp804eh/image/upload/v1561045947/airmodules/Utils/NoImageFound.svg'` |
| translate | `func` |  | `() => {}` |