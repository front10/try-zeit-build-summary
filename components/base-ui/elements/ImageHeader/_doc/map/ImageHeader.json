{"description":"","displayName":"ImageHeader","methods":[{"name":"componentDidUpdate","docblock":null,"modifiers":[],"params":[{"name":"prevProps","type":null}],"returns":null},{"name":"handleLoadImage","docblock":null,"modifiers":[],"params":[],"returns":null},{"name":"onUpdate","docblock":null,"modifiers":[],"params":[{"name":"prevProps","type":null}],"returns":null},{"name":"componentDidMount","docblock":null,"modifiers":[],"params":[],"returns":null},{"name":"componentWillUnmount","docblock":null,"modifiers":[],"params":[],"returns":null},{"name":"listener","docblock":null,"modifiers":[],"params":[],"returns":null},{"name":"addEventListener","docblock":null,"modifiers":[],"params":[],"returns":null},{"name":"removeEventListener","docblock":null,"modifiers":[],"params":[],"returns":null},{"name":"render","docblock":null,"modifiers":[],"params":[],"returns":null}],"props":{"className":{"type":{"name":"string"},"required":false,"description":"Class name of header","defaultValue":{"value":"''","computed":false}},"style":{"type":{"name":"objectOf","value":{"name":"any"}},"required":false,"description":"Css style applied to the component","defaultValue":{"value":"null","computed":false}},"image":{"type":{"name":"string"},"required":false,"description":"URI of the image.","defaultValue":{"value":"''","computed":false}},"alt":{"type":{"name":"string"},"required":false,"description":"alt description","defaultValue":{"value":"null","computed":false}},"imageHeight":{"type":{"name":"string"},"required":false,"description":"Height of Destination Image","defaultValue":{"value":"'25vh'","computed":false}},"imageWidth":{"type":{"name":"string"},"required":false,"description":"Width of Destination Image","defaultValue":{"value":"'100%'","computed":false}},"hoverEffect":{"type":{"name":"enum","value":[{"value":"'grayscale'","computed":false},{"value":"'fadeIn'","computed":false},{"value":"'scale'","computed":false},{"value":"'blur'","computed":false},{"value":"'opacity'","computed":false},{"value":"'sepia'","computed":false},{"value":"'lightly'","computed":false}]},"required":false,"description":"Effect apply of the image","defaultValue":{"value":"null","computed":false}},"focusable":{"type":{"name":"bool"},"required":false,"description":"Enable focusable in this components.","defaultValue":{"value":"true","computed":false}},"children":{"type":{"name":"union","value":[{"name":"arrayOf","value":{"name":"node"}},{"name":"node"}]},"required":false,"description":"Children of the component","defaultValue":{"value":"null","computed":false}}},"filename":"/Users/darielnoel/repositories/front10/airmodules/src/components/elements/ImageHeader/ImageHeader.jsx"}