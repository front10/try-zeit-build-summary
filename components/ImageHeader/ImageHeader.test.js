import React from 'react';
import { render, waitForElement, cleanup } from 'test-utils';
import ImageHeader from './ImageHeader';




test('render all', async () => {
  const { container, getByTestId } = render(
    <ImageHeader
      image="https://www.latam.com/content/dam/LATAM/latam-marca-unica/Home%20LATAM%20USA/Vitrinas%20Desktop/santiago-flights-latam-1500x500"
      imageHeight="30vh"
      imageWidth="90%"
      label="Testing"
      labelSecondary="Testing"
    />
  );
  await waitForElement(() => getByTestId('imageHeader'));
  expect(container).toMatchSnapshot();
});
