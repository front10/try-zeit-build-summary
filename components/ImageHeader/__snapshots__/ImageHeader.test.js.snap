// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render all 1`] = `
<div>
  <div
    class="fr-ImageHeader-container fr-focus fr-h-100"
    data-testid="imageHeader"
    id="fr-image-1557335243965_022747485686284308"
    role="img"
    tabindex="0"
  >
    <div
      class="fr-image-header null fr-focus"
      style="height: 30vh; width: 90%; background-color: lightgray; animation: pulse 1s infinite ease-in-out;"
    />
  </div>
</div>
`;
